### DISCUSSION TITLE
How to prevent user from registering?

### DISCUSSION THREAD

**Discussion by SkYNewZ** [pocketbase/pocketbase#4437] [Q&A] [answered]
Hi! I would like to limit the number of users possible in my application. **Authentication is only possible via OAuth2.** What event can I use to prevent registration if the maximum number has been reached? I'm currently using `app.OnRecordBeforeAuthWithOAuth2Request().PreAdd` and I'm ignoring the case where `e.IsNewRecord` is `false` as I only want this to fire for new users.

However, even after deleting my previously successfully logged in user, `IsNewRecord` still seems `false`.

<details>
<summary>The code</summary>

```go
// bindHookMaxUsersAllowed sets up the hook to check if the maximum number of users is reached.
func bindHookMaxUsersAllowed(app core.App) {
	app.OnRecordBeforeAuthWithOAuth2Request().PreAdd(func(e *core.RecordAuthWithOAuth2Event) error {
		logger := logging.GetLoggerFromApp(app).With(slogx.Component("hooks"), slogx.Hook("max_users"))
		conf := config.FromApp(app)

		logger.Info("checking if maximum number of users is reached")
		if conf.MaxUsers < 0 {
			logger.Info("no max users limit set")
			return nil
		}

		if !e.IsNewRecord {
			return nil // only check for new records
		}

		// check if max users limit is reached
		logger.Warn(fmt.Sprintf("max users limit set to %d, checking if user can signup", conf.MaxUsers))

		var count int
		err := app.Dao().DB().Select("COUNT(id)").From(e.Collection.TableName()).WithContext(e.HttpContext.Request().Context()).Row(&count)
		if err != nil {
			logger.With(slogx.Err(err)).Error("error querying users, cannot check max users")
			return nil // we don't want to block the user from logging in
		}

		if count >= conf.MaxUsers {
			logger.With(slog.Int("max_users", conf.MaxUsers), slog.Int("registered_users", count)).Warn("maximum number of users reached")
			return fmt.Errorf("maximum number of users allowed reached: %d (%w)", conf.MaxUsers, hook.StopPropagation)
		}

		return nil
	})
}
```

Also moving to this does not solve the issue as `Record` is `nil`.
```go
if !e.Record.IsNew() {
	return nil // only check for new records
}
```

</details>

Could be linked to https://github.com/pocketbase/pocketbase/discussions/4357#discussioncomment-8485292 I suppose.

**Comment by ganigeorgiev** [comment, marked as answer]
> ... `IsNewRecord` still seems `false`.

This looks more like a bug or an oversight on my part when I added the event field.

At the moment `IsNewRecord` will report a correct value but only in the After hook.

With the planned refactoring I'll consider fixing it to be consistent for both events.

In any case, as an alternative to the field in the before hook is to check whether `e.Record` is `nil`.

**Reply by SkYNewZ** [reply]
Indeed the `IsNewRecord` does not seems to be filled here. https://github.com/pocketbase/pocketbase/blob/f414e70ffa2255d0acfd7b2832def82c18137ea2/apis/record_auth.go#L252

I can check the newness of the record if `e.Record` is `nil`. Thanks.

<details>
<summary>This is the final code</summary>

```go
func bindHookMaxUsersAllowed(app core.App) {
	app.OnRecordBeforeAuthWithOAuth2Request().PreAdd(func(e *core.RecordAuthWithOAuth2Event) error {
		logger := logging.GetLoggerFromApp(app).With(slogx.Component("hooks"), slogx.Hook("max_users"))

		conf := config.FromApp(app)
		if conf.MaxUsers < 0 {
			logger.Debug("no max users limit set")
			return nil
		}

		// nil Record means that it will try to automatically create one (if it can)
		if e.Record != nil {
			return nil
		}

		// check if max users limit is reached
		logger.Warn(fmt.Sprintf("max users limit set to %d, checking if user can signup", conf.MaxUsers))

		var count int
		err := app.Dao().DB().Select("COUNT(id)").From(e.Collection.Name).WithContext(e.HttpContext.Request().Context()).Row(&count)
		if err != nil {
			logger.With(slogx.Err(err)).Error("error querying users, cannot check max users")
			return nil // we don't want to block the user from logging in
		}

		if count >= conf.MaxUsers {
			logger.With(slog.Int("max_users", conf.MaxUsers), slog.Int("registered_users", count)).Warn("maximum number of users reached")
			return apis.NewForbiddenError("Maximum number of users reached. Please contact administrator.", nil)
		}

		return nil
	})
}
```
</details>

**Reply by ganigeorgiev** [reply]
I'll consider fixing it either with the next v0.22.1 release or with the refactoring in v0.23.0 (_technically `IsNewRecord` should be converted to a getter and not an editable struct field because users can assign their own custom `e.Record` to link and the value could become again out of sync, or at least maybe rename it to indicate that this was only the initial state_). Will think on it tomorrow.

**Reply by SkYNewZ** [reply]
Thank you for your reply. There's no need to rush - it works just fine that way! 😄 Thank you for being so available.
